slackapi/deno-slack-hooks
# Add install-update hook to update dependencies

###  Summary

This pull request introduces an `install-update` hook. When executed, this hook runs a check for updates, updates the dependency file(s) (`import_map.json`, `slack.json`) with the latest release versions, and then runs the `build` hook as the means to see the dependency updates cached.

https://user-images.githubusercontent.com/7788242/179641508-4fa72443-f280-47df-a91a-919e3fcc0aa6.mov

### Implementation Notes

#### Familiar Structures
Relying on the prior art of the `check-update` hook, the code here should look very familiar, right down to the data structures:

```
interface InstallUpdateResponse {
  name: string;
  updates: Update[];
  error?: {
    message: string;
  } | null;
}

export interface Update {
  name: string;
  previous: string;
  installed: string;
  error?: {
    message: string;
  } | null;
}
```

At first glance, some of these properties might seem superfluous, but they're crucial in providing a solid experience on the CLI end of things. 

Rather than returning a mere `status: "ok"` with an error, when applicable, bundling the updates with individual information about the outcome of the installation attempt allows for the CLI to reliably display which installations were successful vs. not, and, if so, which version the user should find in their updated project, as demonstrated in the error state video below:

https://user-images.githubusercontent.com/7788242/179640616-d0a6060f-61b1-4e0c-88ef-c0fc98260546.mov

### Testing Steps
The full experience requires the CLI to work in concert with this update. However, to test out the JSON that's returned, in a new project: 

1. Check out this branch locally

2. In a project, bump the versions in your `import_map.json` and/or `slack.json` to something less than what is currently there.

3. Run the following command (which is what will normally be in the `install-update` hook):

```
deno run -q --config=deno.jsonc --allow-run --allow-read --allow-write --allow-net file:///Users/<YOUR_PATH_HERE>/deno-slack-hooks/src/install_update.ts
```

4. Verify that the JSON output contains the appropriate updates, per the changes you manually made before running the command.

### Requirements (place an `x` in each `[ ]`)
